File: frontend/src/utils/paymentValidation.js
/**
 * Validation côté client du formulaire de paiement (miroir du backend).
 */

export function validateCardFormat(cardNumber) {
  const digits = cardNumber.replace(/\D/g, '');
  return /^\d{13,19}$/.test(digits);
}

export function validateEmail(email) {
  return /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/.test(email.trim());
}

export function validateExpiry(expiry) {
  const match = expiry.trim().match(/^(0[1-9]|1[0-2])\s*\/\s*(\d{2}|\d{4})$/);
  if (!match) return false;
  const month = parseInt(match[1], 10);
  const yearRaw = match[2];
  const year = yearRaw.length === 4 ? parseInt(yearRaw, 10) : 2000 + parseInt(yearRaw, 10);
  const now = new Date();
  const expEnd = new Date(year, month, 0);
  return expEnd >= new Date(now.getFullYear(), now.getMonth(), 1);
}

export function formatCardNumber(value) {
  const digits = value.replace(/\D/g, '').slice(0, 19);
  return digits.replace(/(\d{4})(?=\d)/g, '$1 ').trim();
}

export function formatExpiry(value) {
  const digits = value.replace(/\D/g, '').slice(0, 4);
  if (digits.length <= 2) return digits;
  return `${digits.slice(0, 2)}/${digits.slice(2)}`;
}

export function validateBillingForm(data, plan, lang = 'fr') {
  const t = (fr, ar) => (lang === 'ar' ? ar : fr);

  if (!data.prenom?.trim() || data.prenom.trim().length < 2) {
    return t('Le prénom doit contenir au moins 2 caractères.', 'يجب أن يحتوي الاسم الشخصي على حرفين على الأقل.');
  }
  if (!data.nom?.trim() || data.nom.trim().length < 2) {
    return t('Le nom doit contenir au moins 2 caractères.', 'يجب أن يحتوي اسم العائلة على حرفين على الأقل.');
  }
  if (!validateEmail(data.email || '')) {
    return t('Adresse e-mail invalide.', 'البريد الإلكتروني غير صالح.');
  }

  if (plan === 'free') return null;

  const digits = (data.carte || '').replace(/\D/g, '');
  if (!validateCardFormat(digits)) {
    return t('Numéro de carte invalide (13 à 19 chiffres).', 'رقم البطاقة غير صالح (13 إلى 19 رقمًا).');
  }
  if (!validateExpiry(data.expiry || '')) {
    return t('Date d\'expiration invalide ou carte expirée (MM/AA).', 'تاريخ انتهاء الصلاحية غير صالح (MM/AA).');
  }
  if (!/^\d{3,4}$/.test((data.cvc || '').trim())) {
    return t('CVC invalide (3 ou 4 chiffres).', 'رمز CVC غير صالح (3 أو 4 أرقام).');
  }

  return null;
}

File: backend/app/services/payment_validation.py
import re
from datetime import datetime


def validate_card_format(card_number: str) -> bool:
    """
    Valide le format d'un numéro de carte bancaire.
    Accepte 13 à 19 chiffres (Visa, Mastercard, Amex, etc.) sans exiger Luhn.
    """
    digits = normalize_card_digits(card_number)
    return bool(re.match(r"^\d{13,19}$", digits))


def validate_email(email: str) -> bool:
    pattern = r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$"
    return bool(re.match(pattern, email.strip()))


def validate_card_expiry(expiry: str) -> bool:
    """Valide le format MM/YY ou MM/YYYY et vérifie que la carte n'est pas expirée."""
    expiry = expiry.strip()
    match = re.match(r"^(0[1-9]|1[0-2])\s*/\s*(\d{2}|\d{4})$", expiry)
    if not match:
        return False
    month = int(match.group(1))
    year_raw = match.group(2)
    year = int(year_raw) if len(year_raw) == 4 else 2000 + int(year_raw)
    now = datetime.utcnow()
    if year < now.year or (year == now.year and month < now.month):
        return False
    return True


def validate_cvc(cvc: str) -> bool:
    return bool(re.match(r"^\d{3,4}$", cvc.strip()))


def normalize_card_digits(card: str) -> str:
    return re.sub(r"\D", "", card)


def card_last4(card: str) -> str:
    digits = normalize_card_digits(card)
    return digits[-4:] if len(digits) >= 4 else digits


def validate_payment_form(
    *,
    forfait: str,
    client_nom: str,
    client_prenom: str,
    client_email: str,
    numero_carte: str = "",
    card_expiry: str = "",
    card_cvc: str = "",
) -> tuple[bool, str]:
    """
    Valide le formulaire de réservation.
    Retourne (True, "") ou (False, message_d_erreur).
    """
    if forfait not in ("free", "pro", "max"):
        return False, "Forfait invalide."

    if len(client_prenom.strip()) < 2:
        return False, "Le prénom doit contenir au moins 2 caractères."
    if len(client_nom.strip()) < 2:
        return False, "Le nom doit contenir au moins 2 caractères."
    if not validate_email(client_email):
        return False, "Adresse e-mail invalide."

    if forfait == "free":
        return True, ""

    digits = normalize_card_digits(numero_carte)
    if not validate_card_format(digits):
        return False, "Numéro de carte invalide (13 à 19 chiffres requis)."
    if not validate_card_expiry(card_expiry):
        return False, "Date d'expiration invalide ou carte expirée (format MM/AA)."
    if not validate_cvc(card_cvc):
        return False, "CVC invalide (3 ou 4 chiffres)."

    return True, ""
